kernel: arch interface housekeeping

tyrantlizard77 opened the pull request:
To prepare for the centralized arch interface in #19584 some housekeeping is in order, in that all arches must define these APIs in the same way.

APIs which are defined as inline functions, but implemented non-inline, or vice versa, are now consistent across all arches and in conformance with the arch header.

APIs which are defined as inline functions, but implemented as macros, are now implemented as inline functions.

x86_64 needed a little more work as arch/cpu.h was pulling in kernel_arch_func.h. which shouldn't be done.

Commit:
posix: expose z_arch_irq_lock/unlock as inlines

The specification for these arch APIs is to have them inline,
and the bodies were just oneliners calling another function
anyway.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arches: fix z_arch_is_in_isr() defintion

For some reason these were implemented as macros when they
should be inline functions.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arm: move z_arch_switch_to_main_thread to C code

There's no compelling reason why this should be inline unlike all
other arches, it's a large function, called exactly once.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
drivers: hpet: fix includes

IRQ APIs were being used without pulling in the proper header.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
xtensa: fix z_arch_switch()

This was in the wrong header and declared as a macro instead
of an inline function.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
arches: fix z_arch_k_cycle_get_32() definitions

These need to all be inline functions and not macros.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
intel64: add inline definition of z_arch_switch()

This just calls the assembly code. Needed to be
inline per the architecture API.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
x86_64: fix arch headers

arch/cpu.h and kernel_arch_func.h are expected to define different
functions, per the architecture interface.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
sys_io: pull in toolchain.h

Needed for ALWAYS_INLINE.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel_includes: add irq.h

A lot of code assumes kernel.h pulls in irq_lock(),
guarantee this.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

aescolar commented on arch/posix/include/posix_soc_if.h:
```diff
@@ -35,6 +33,16 @@ void posix_irq_full_unlock(void);
 int  posix_get_current_irq(void);
 /* irq_offload() from irq_offload.h must also be defined by the SOC or board */
 
+static inline unsigned int z_arch_irq_lock(void)
```
I would not define these here, but maybe in include/arch/posix/arch.h 
(both for consistency with other archs, and because this header is meant to define the interface a posix soc provides to the posix arch)

If you are doing this change, I would then also be consistent and do the same for the other z_arch_irq* functions here. I think it makes sense if I send a quick PR and then if you want you can take that commit into this one.

aescolar replied:
@andrewboie I just added an extra patch on top of this branch in here:
https://github.com/aescolar/zephyr/commits/pa_hide_internal_if 
https://github.com/aescolar/zephyr/commit/92671a6a03395ca55fb084da90c33cad596c36f9
Feel free to take it in if you agree. Or tell how you would prefer to handle it otherwise.

andrewboie replied:
I pulled your patch into the series.

aescolar replied:
Thanks @andrewboie . This new commit is indeed going a bit further moving things around, but as this arch API changes are also causing extra merge work for off-tree POSIX boards I thought it better to take them in one go, and put those inlines in a more proper place. It should also simplify a bit future arch interface cleanup PRs (so you dont need to dig thru the posix SOC and boards files)

aescolar committed:
arch: posix: isolate arch-soc/board IF from kernel-arch IF

The POSIX ARCH delegates some of the tasks which normally
are taken care of by the ARCH to the SOC or BOARD levels.

To avoid changes in the kernel-arch IF propagating into
the arch-soc and arch-board interfaces (which would break
off-tree posix boards) isolate them.

Also move arch inlined functions into the arch.h header,
and out from the headers which specify the  posix arch-soc
and arch-board interfaces.

Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>

andrewboie commented:
Rebased and added @aescolar's patch to the series.

ioannisg reviewed (commented):
If I understood correctly, z_arch_k_cycle_get_32() is inlined as z_timer_cycle_get_32() for _all_ ARCHes, except, perhaps the x86_64. So I wonder if we could have this as common and clean this up from all architecture inlines?

aescolar reviewed (approved):
Ok for POSIX arch/soc/boards related changes, no claim done for other arches

ioannisg reviewed (approved):
ARM changes look good; happy that functions are moved to C files.

ioannisg commented:
@ruuddw @nategraff-sifive perhaps a quick look into ARC and RISC-V?

vonhust reviewed (approved):
OK for ARC

ioannisg commented:
> 
> 
> > If I understood correctly, z_arch_k_cycle_get_32() is inlined as z_timer_cycle_get_32() for all ARCHes, except, perhaps the x86_64. So I wonder if we could have this as common and clean this up from all architecture inlines?
> 
> I'm not going to do this in this PR. It's not in the scope of what this PR is trying to do. I suggest filing a bug on it, there may be reasons why this is left up to the arch port and not hard-coded to the timer.

Will do, although that's not a bug; perhaps a question.

andyross commented:
> there may be reasons why this is left up to the arch port and not hard-coded to the timer.

I wouldn't say "reasons", but in principle there's no reason for the cycle counter to be tied to the timer counter.  For platforms like nRF, it actually would be useful to have a MHz SysTick-based cycle counter and a kHz RTC at the same time, though that's not the way it's historically worked.  Leaving the definition to the arch/soc layers at least allows that sort of change compatibly in the future.
